pkg/calc/node_counter.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package calc
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+)
+
+func NewNodeCounter(sink api.SyncerCallbacks) *NodeCounter {
+	return &NodeCounter{
+		sink: sink,
+	}
+}
+
+type NodeCounter struct {
+	sync.Mutex
+	sink      api.SyncerCallbacks
+	inSync    bool
+	nodeCount int
+}
+
+func (c *NodeCounter) OnStatusUpdated(status api.SyncStatus) {
+	if status == api.InSync {
+		c.Lock()
+		c.inSync = true
+		c.Unlock()
+	}
+	c.sink.OnStatusUpdated(status)
+}
+
+func (c *NodeCounter) OnUpdates(updates []api.Update) {
+	for _, update := range updates {
+		switch update.Key.(type) {
+		case model.NodeKey:
+			switch update.UpdateType {
+			case api.UpdateTypeKVNew:
```

Comment:
This is 100% correct according to the syncer API but I'm a bit wary of refcounting in case there's a subtle bug in libcalico-go's setting of the update type.

Given there are at most a few thousand nodes, I think I'd use a set and then add/remove the nodenames from the set (and then calculate the nod count from the size of the set).  WDYT?